packages/core/strapi/lib/services/entity-service/types/params/index.d.ts

```diff
@@ -1,22 +1,41 @@
-import type { Common } from '@strapi/strapi';
-
+import type { Common, EntityService } from '@strapi/strapi';
+// Params
 import type * as Sort from './sort';
 import type * as Pagination from './pagination';
 import type * as Fields from './fields';
 import type * as Filters from './filters';
 import type * as Populate from './populate';
 import type * as PublicationState from './publication-state';
 import type * as Data from './data';
-
+// Utils
 import type * as Attribute from './attributes';
 
-export type For<TSchemaUID extends Common.UID.Schema> = {
+type Read<TSchemaUID extends Common.UID.Schema> = {
   sort?: Sort.Any<TSchemaUID>;
   fields?: Fields.Any<TSchemaUID>;
   filters?: Filters.Any<TSchemaUID>;
   populate?: Populate.Any<TSchemaUID>;
-  data?: Data.Input<TSchemaUID>;
 } & Pagination.Any &
-  PublicationState.For<TSchemaUID>;
+  PublicationState.For<TSchemaUID> &
+  EntityService.GetPluginReadParams<TSchemaUID>;
+
+type Write<TSchemaUID extends Common.UID.Schema> = {
+  data?: Data.Input<TSchemaUID>;
+  files?: unknown; // TODO
+} & EntityService.GetPluginWriteParams<TSchemaUID>;
+
+type ReadWrite<TSchemaUID extends Common.UID.Schema> = Read<TSchemaUID> & Write<TSchemaUID>;
```

Comment:
I have another "do we update the documentation or the code?" question here.

Create/update/delete only document that `populate` and `fields` are available, though I wouldn't be surprised if `sort` and maybe even `filters` also worked.  I don't know if we want to include all of the Read params or just those. Though I suppose they could just pick/omit what they need. What do you think?

Reply:
I am just thinking why do we even need Read and Write ?  as mentioned by Ben the methods can contain a different combination of params options so we would start having to omit stuff.
why don't we leave the definition of the param object to when we define the methods themselves ?

Reply:
True, it might be simplest to just have discrete params and each method can pick and choose what they support.

Reply:
We can still `pick` or `omit` from the Read/Write/ReadWrite types. I like the idea of having types to group common patterns & usages though.

> why don't we leave the definition of the param object to when we define the methods themselves 

My 2 main concerns would be verbosity & maintainability

Having to pick every param again & again in each method is very verbose and I'm not sure it would be very scalable if we add new ones in the future.

As it stands right now, we have 2 different main usages:

- Params used in queries to write in DB (data/files)
- Params used in queries to read from DB (all the rest)
(+ we have dynamic parameters such as locale that comes from i18n)

On top of that some methods only accept some params from the main Read params (often fields & populate, which are called "selection params"), hence the types created in the other PR (findOne)

Finally, we do need to have this big Read type to use in other places such as the Populate object notation (nested populate)

Reply:
I can see either way. I think this way (grouped together) is best as long as we intend to always support an identical set of parameters in every method, and only have rare exceptions to it, in which case we can pick/omit.

If we have many exceptions though, I can see it getting messy.

But you did bring up a good point, we might want to break out Selection (fields+populate) so it's easier to use just those two.